### DISCUSSION TITLE
Improved martensite variant indexing?

### DISCUSSION THREAD

**Discussion by nyyssont** [mtex-toolbox/mtex#828] [Ask Anything]
When martensite transforms from austenite, 24 crystallographically distinct variants may form. In low-carbon martensite, the variant forms as what is generally called a "lath", referring to the physical shape of the formed variant.

The 011 plane of each martensitic lath is close to parallel with one of the austenite planes (1,1,1), (1,-1,1), (-1,1,1), (1,1,-1). The laths that all share the same (011) to {111} parallelism are defined as being in the same "packet". The group is called a packet because the laths in the same packet tend to form next to each other, thereby forming a physical packet of laths.

The packet indexing has been incorporated into MTEX and can be accessed using the calcChildVariant command. It works reliably.

Within a packet, each martensitic lath has a parallel [-1,-1,1] or a [-1,1,-1] direction to one of the <101> directions of austenite that lie on the parallel {111} plane. For the first packet (111) in austenite is close to parallel with (011) in martensite. This means the [-101], [1-10] and [01-1] directions in austenite are parallel with either [-1,-1,1] or a [-1,1,-1] in martensite.

Morito et al. [1] have defined the order of the variants in a packet as follows:
![variant_indexing_Morito_et_al](https://user-images.githubusercontent.com/10060198/109972837-04d91000-7cf8-11eb-9370-2f296c18a16b.png)

This order is followed in a number of publications. I would like for the variant indexing given by calcChildVariant in MTEX to follow this definition, at least to the extent that I could say "OK, this is variant 2 as determined by Morito and I can expect for it to have this type of relationship with variant 5" and so on.

I seem to have a lot of trouble with the parallel directions, however. MTEX says that for each variant, it is always a [-1,-1,1] direction in martensite that is parallel with the <101> type direction in austenite. The figure below shows my attempt to plot the martensite [-1,-1,1] directions on top of the austenite <101> directions for the six variants in packet 1, as identified by calcChildVariant.

![planes_directions](https://user-images.githubusercontent.com/10060198/109977749-6e0f5200-7cfd-11eb-9d3f-ea4f50afbc5c.png)

Can anyone explain what is going wrong here? Does anyone have any suggestion on a sensible way to implement Morito-style variant indexing? Are there other ways of indexing that would make sense and are more MTEX-friendly? All remarks and conversation on the topic are welcome.

[1] S. Morito, H. Tanaka, R. Konishi, T. Furuhara, T. Maki Acta Materialia, https://doi.org/10.1016/S1359-6454(02)00577-3.

The code to generate the figure is below:

```
plotx2east

%Define some phases:
cs_bcc = crystalSymmetry('m-3m', [2.9 2.9 2.9], 'mineral', 'Iron bcc (old)');
cs_fcc = crystalSymmetry( 'm-3m', [3.7 3.7 3.7], 'mineral', 'Iron fcc');

% Define the parallel planes and directions (K-S):
KS = orientation.map(Miller(1,1,1,cs_fcc),Miller(0,1,1,cs_bcc),...
      Miller(-1,0,1,cs_fcc),Miller(-1,-1,1,cs_bcc));

% Use Greninger-Trojano, define as "close to K-S"
p2c = orientation.GreningerTrojano(cs_fcc,cs_bcc);
p2c = p2c.project2FundamentalRegion(KS)

oripar = orientation.cube(cs_fcc); % parent orientation
childOri = variants(p2c,oripar); % child orientations

%packet ids
h_p_plane = Miller({1,1,1},{1,-1,1},{-1,1,1},{1,1,-1},p2c.CS);
h_c_plane = Miller(0,1,1,p2c.SS); % child 011 plane

omega = dot(variants(p2c,h_p_plane),h_c_plane);
[~,packetId] = max(omega,[],2);

h_p_dir = Miller({-1,0,1},{1,-1,0},{0,1,-1},cs_fcc,'uvw') % packet 1 parent directions
h_c_dir = Miller(-1,-1,1,p2c.SS,'uvw'); % child -1-11 direction

figure
%parent directions in packet 1 (on parent plane 111)
plot(oripar*(Miller({-1,0,1},{1,-1,0},{0,1,-1},cs_fcc,'uvw')),'labeled',...
    'MarkerSize',10,'MarkerFaceColor','red','color','red')
hold on
% 111 parent plane and direction
plot(oripar*h_p_plane(1),'plane','linecolor','r','linewidth',2,'add2all')
annotate(oripar*h_p_plane(1),...
    'marker','o','MarkerSize',30,'MarkerFaceColor','none','MarkerEdgeColor','red','linewidth',2,...
    'label',char(h_p_plane(1),'cell','Latex'),'color','red')

% child [-1 -1 1] directions for first packet, variant 1 marked
plot(childOri(1)*h_c_dir,...
    'marker','o','MarkerSize',15,'MarkerFaceColor','none','MarkerEdgeColor','black','add2all')
text(childOri(1)*h_c_dir,'Variant 1','horizontalAlignment','left','VerticalAlignment','bottom','add2all')
plot(childOri(packetId == 1)*h_c_dir,...
    'marker','o','MarkerSize',10,'MarkerFaceColor','green','MarkerEdgeColor','black',...
    'DisplayName',[char(h_c_dir,'Latex') ' $\alpha$'],'color','green','add2all')

%child 011 plane normal:
plot(childOri(packetId == 1)*h_c_plane,...
    'marker','o','MarkerSize',10,'MarkerFaceColor','none','MarkerEdgeColor','black',...
    'DisplayName',[char(h_c_plane,'Latex') ' $\alpha$'],'add2all')

legend('Interpreter','latex','FontSize',20,'EdgeColor','none','location','west')
```

**Comment by nyyssont** [comment]
Previously, I've ordered the variants manually:

```
varId = calcChildVariant(parentOri,childOri,p2c);

ord = [1 3 5 21 23 19 11 7 9 16 14 18 24 22 20 4 2 6 13 15 17 8 12 10];

varId_ordered = ord(varId);
```

As a reference, I've used the 100 pole figure for martensite variants presented in Figure 1 in [1]. Doing this results in the variant order shown below (same as in [1]):

![variant_indexing_Morito_et_al_2](https://user-images.githubusercontent.com/10060198/110239557-dfdadc00-7f47-11eb-9b14-57541fdf3186.png)

This works, but it would be nice to end up with the same variant ordering using MTEX operations, so (for example) a change in the order of symmetry operators wouldn't spoil things.

In the code snippet below is an attempt to sort the variants using MTEX operations:

<details><summary><b>Click to view script 1/2: Variant ordering</b></summary>
<p>

```matlab
plotx2east

%Define prior and product phases:
cs_bcc = crystalSymmetry('m-3m', [2.9 2.9 2.9], 'mineral', 'Iron bcc (old)');
cs_fcc = crystalSymmetry( 'm-3m', [3.7 3.7 3.7], 'mineral', 'Iron fcc');

KS = orientation.map(Miller(1,1,1,cs_fcc),Miller(0,1,1,cs_bcc),...
      Miller(-1,0,1,cs_fcc),Miller(-1,-1,1,cs_bcc));

% Use Greninger-Trojano OR, define as "close to K-S"
p2c = orientation.GreningerTrojano(cs_fcc,cs_bcc);
p2c = p2c.project2FundamentalRegion(KS)

oripar = orientation.cube(cs_fcc); % parent orientation
childOri = variants(p2c,oripar); % child orientations

%% planes and directions

h_p_plane = Miller({1,1,1},{1,-1,1},{-1,1,1},{1,1,-1},p2c.CS); %this will determine packet order
h_c_plane = Miller(0,1,1,p2c.SS); % child 011 plane

h_p_dir = Miller({-1,0,1},{0,1,-1},{1,-1,0},cs_fcc,'uvw') % this will determine variant order
h_c_dir = Miller(-1,-1,1,p2c.SS,'uvw'); % child -1-11 direction

%% packet ids:
omega = dot(variants(p2c,h_p_plane),h_c_plane);
[~,packetId] = max(omega,[],2);

%% rotate all child orientations to packet 1 for variant ordering:
rot_1 = rotation.map(h_p_plane(2),h_p_plane(1),Miller(1,0,-1,cs_fcc,'uvw'),h_p_dir(1));
rot_2 = rotation.map(h_p_plane(3),h_p_plane(1),Miller(0,-1,1,cs_fcc,'uvw'),h_p_dir(1));
rot_3 = rotation.map(h_p_plane(4),h_p_plane(1),Miller(-1,1,0,cs_fcc,'uvw'),h_p_dir(1));

new_childs = childOri;
new_childs(packetId == 2) = rot_1*childOri(packetId == 2)
new_childs(packetId == 3) = rot_2*childOri(packetId == 3)
new_childs(packetId == 4) = rot_3*childOri(packetId == 4)

%% reorder variant list
varord = zeros(24,1); %reordered variant list

% Separate 1,3,5 and 2,4,6:
ind_1 = angle(new_childs*h_c_plane,oripar*h_p_plane(1))/degree < 1
ind_2 = angle(new_childs*h_c_plane,oripar*h_p_plane(1))/degree > 1

omega = dot(h_p_dir,new_childs*h_c_dir);
[~,varId] = max(omega,[],2);

varord(ind_1) = varId(ind_1)*2-1 %variants 1,3,5
varord(ind_2) = varId(ind_2)*2 %variants 2,4,6

varord(packetId == 2) = varord(packetId == 2)+6 % packet 2
varord(packetId == 3) = varord(packetId == 3)+12 % packet 3
varord(packetId == 4) = varord(packetId == 4)+18 % packet 4

plotPDF(childOri,Miller(0,0,1,cs_bcc))
annotate(childOri*Miller(0,0,1,cs_bcc),...
    'marker','o','MarkerSize',10,'MarkerFaceColor','none','MarkerEdgeColor','black',...
    'label',varord,'antipodal','FontSize',20)
```

</p>
</details>

This seems to work, although I'm a bit uncertain about having to sort the variants 1,3,5 and 2,4,6 based on whether the 011 martensite plane normal is pointing to 111 austenite plane normal or not. Below is a comparison on the variant numbering with Figure 1 from [1]:

![Miyamoto_variant_indexing](https://user-images.githubusercontent.com/10060198/110241207-d9506280-7f4f-11eb-8118-5577c7c50f31.png)

Any comments or attempts to make the snippet cleaner, lighter or easier to follow would be appreciated.

In the below code, the variant order determined by the snippet above has been used to color the variants in the demo martensite dataset after reconstruction.

<details><summary><b>Click to view script 2/2: Variant coloring</b></summary>
<p>

```matlab
% load the data
mtexdata martensite 
plotx2east

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize < 3)).rotations = quaternion.nan;
ebsd(grains(grains.grainSize < 3)).phase = 0;
ebsd(grains(grains.grainSize < 3)).grainId = 0;
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
%%
job = parentGrainReconstructor(ebsd,grains) % define job
job.calcParent2Child; % get representative OR

%%
job.calcGBVotes % compute votes from child grain boundaries
job.calcParentFromVote('minVotes',4)

%%
for l = 1:2
job.calcGBVotes('noC2C') % compute votes ONLY from parent-child boundaries;
job.calcParentFromVote('minFit',5*degree) %compute parent orientations from votes
end

%% Variant ordering with snippet from above:
varord = [1 3 5 21 23 19 11 7 9 16 14 18 24 22 20 4 2 6 13 15 17 8 12 10]';

%%

parentEBSD = job.calcParentEBSD;
[grains_aus,parentEBSD('Iron fcc').grainId] = calcGrains(parentEBSD('Iron fcc'),'angle',3*degree)
ebsd.grainId = parentEBSD.grainId

varId = calcChildVariant(grains_aus(parentEBSD('Iron fcc').grainId).meanOrientation.project2FundamentalRegion,ebsd(parentEBSD('Iron fcc').id).orientations,job.p2c);

figure
plot(parentEBSD('Iron fcc'),varord(varId))
colormap jet
hold on
plot(grains.boundary,'linewidth',2)
plot(grains_aus.boundary,'linewidth',4,'linecolor','black')
plot(grains_aus.boundary,'linewidth',1,'linecolor','cyan')
```

</p>
</details>

The jet colormap handily separates into four distinct color schemes for the variants. The blue variants belong to packet 1, the blue-green to packet 2, the yellow to packet 3 and the red to packet 4.

![variant_coloring](https://user-images.githubusercontent.com/10060198/110241658-25040b80-7f52-11eb-8aaa-15ce843d4416.jpg)

In the image, the initial grain boundaries are shown in black and the PAG boundaries in cyan. Careful inspection will show that inside the initial grain boundaries, almost every grain is subdivided into several variants. In each case, the variants have the low-angle V1/V4 type relationship, as determined by Morito et al [2].

Any comments? Should something like this be included as an option in calcChildVariant:

```
[varId, packetId] = calcChildVariant(parentOri,childOri,p2c,'Morito')
```

Or would it be better as its own function, perhaps? Or is imposing this type of ordering something that should be kept out of MTEX?

[1] G. Miyamoto, N. Iwata, N. Takayama , T. Furuhara, Acta Materialia, https://doi.org/10.1016/j.actamat.2011.11.018
[2] S. Morito, H. Tanaka, R. Konishi, T. Furuhara, T. Maki, Acta Materialia, https://doi.org/10.1016/S1359-6454(02)00577-3.

**Reply by Tijmenvermeij** [reply]
Dear Tuomo,

I don't have very helpful comments (yet), but thanks for outlining these issues. 
I do agree it would be preferable if "pure" Mtex operations could be used to retrieve the "Morito style" variant numbering. A 'Morito' option in the calcChildVariant would indeed be useful.

I will try to think a bit about these problems and hopefully have some more useful input later... :)

Best regards,
Tijmen Vermeij

**Reply by ralfHielscher** [reply]
I have also looked into this problem and

1. improved on round2Miller (9ebbca3e361243b35623a505c95855361d989818) - just try ```round2Miller(p2c)```
2. came up with the following ordering algorithm
```matlab
%% get default variants

csParent = crystalSymmetry('432','mineral','child');
csChild = crystalSymmetry('432','mineral','child');

p2c = orientation.KurdjumovSachs(csParent,csChild);
%p2c = orientation.GreningerTrojano(csParent,csChild);
%p2c = orientation.NishiyamaWassermann(csParent,csChild);

p2cV = p2c.variants;

round2Miller(p2cV)

%% determine child symmetry such that (011) is parallel to (XXX)
% with only one negative X 

csParent = p2c.CS;
csChild = p2c.SS;

nParent = Miller({1 1 1},{1 -1 1},{-1 1 1},{1 1 -1},csParent);

d = angle_outer(inv(p2cV) * p2c.SS * Miller(0,1,1,csChild) ,nParent,'noSymmetry');
d = min(d,[],2);
d = reshape(d,length(p2cV),numSym(p2c.SS));

[~,id] = min(d,[],2);

symRot = p2c.SS.rot;
p2cV = inv(symRot(id)) .* p2cV(:);


round2Miller(p2cV)

%% transform to packet 1

dParent = Miller({-1 0 1},{0 1 -1},{1 -1 0},csParent);

% determine packetId
[~,packetId] = min(reshape(angle(p2cV * nParent, Miller(0,1,1,csChild)),[],length(nParent)),[],2);

rot2p1 = orientation.map(nParent(1),nParent, ...
  dParent(1),Miller({-1 0 1},{1 0 -1},{0 -1 1},{-1 1 0},csParent));

p2cp1 = p2cV .* rot2p1(packetId);

round2Miller(p2cp1)

%% determine lathId in packet 1

dChild = Miller({-1 -1 1},{-1 1 -1},csChild);

d = reshape(angle_outer(inv(p2cp1)*dChild,dParent,'noSymmetry'),[],length(dParent)*length(dChild));

% the lath id
[~,lathId] = min(d,[],2);

% the variant Id
vId = 6*(packetId-1) + lathId;

% new order we get from reordering the variant Id
[~,newOrder] = sort(vId);

round2Miller(p2cV(newOrder))

%% check packet Id

[~,p] = calcVariantId(orientation.id(csParent),...
  orientation.id(csParent)*inv(p2cV(newOrder)),p2c)
```
Ralf.

**Comment by Tijmenvermeij** [comment]
A first observation of mine: in the polefigure plot of your first post, the (011) planes that are "not parallel with (111)" just have an opposite direction to the "correct (011) planes" (can be visualized by plotting everything on the upper part of the polefigure using 'antipodal' option for the first plotting command in the script). So all the (011) planes are actually parallel to the (111) plane, but some of the "normal vectors" just have inverted directions. Could this be related to the more general problem of the -11-1 vs -1-11 directions...?

**Reply by nyyssont** [reply]
You're right there. I feel like I was trying to say what you wrote now but I got confused :)
